cmd/bucket-handlers.go

```diff
@@ -1174,6 +1214,84 @@ func (api objectAPIHandlers) PostPolicyBucketHandler(w http.ResponseWriter, r *h
 		}
 	}
 
+	if len(fanOutReq) > 0 {
+		concurrentSize := 100
+		if runtime.GOMAXPROCS(0) < concurrentSize {
+			concurrentSize = runtime.GOMAXPROCS(0)
+		}
+
+		fanOutResp := make([]minio.PutObjectFanOutResponse, 0, len(fanOutReq))
+		for {
+			reader.(io.Seeker).Seek(0, 0) // seek reader back to original position for each iteration.
+
+			var objInfos []ObjectInfo
+			var errs []error
+
+			var done bool
+			if concurrentSize < len(fanOutReq) {
+				objInfos, errs = fanOutPutObject(ctx, bucket, objectAPI, fanOutReq, pReader)
+				done = true
+			} else {
+				objInfos, errs = fanOutPutObject(ctx, bucket, objectAPI, fanOutReq[:concurrentSize], pReader)
+				fanOutReq = fanOutReq[concurrentSize:]
+			}
+
+			for i, objInfo := range objInfos {
+				if errs[i] != nil {
+					fanOutResp = append(fanOutResp, minio.PutObjectFanOutResponse{
+						Key:   objInfo.Name,
+						Error: errs[i],
+					})
+					continue
+				}
+
+				fanOutResp = append(fanOutResp, minio.PutObjectFanOutResponse{
+					Key:          objInfo.Name,
+					ETag:         objInfo.ETag,
+					VersionID:    objInfo.VersionID,
+					LastModified: &objInfo.ModTime,
+				})
+
+				// Notify object created event.
+				defer sendEvent(eventArgs{
```

Comment:
This will potentially defer thousands of calls.

Since we are blocking on them anyway, we should probably just remove the defer - it will also allow senders to catch up if we are doing multiple loops.

Reply:
I think I will perhaps do this as a bulk call at once instead of many events.

Reply:
Yeah, probably a better approach. Maybe spawn a goroutine with a waitgroup, so it will send while the subsequent loops are running (if multiple), and we only wait once all have completed.